libwacom/libwacom-database.c

```diff
@@ -279,8 +323,88 @@ parse_stylus_id(const char *str, WacomStylusId *id)
 	return ret;
 }
 
+static char *
+string_or_fallback (GKeyFile *keyfile, const char *group,
+		    const char *key, const char *fallback) {
+	g_autoptr(GError) local_error = NULL;
+	if (g_key_file_has_key(keyfile, group, key, &local_error))
+		return g_key_file_get_string(keyfile, group, key, NULL);
+
+	return g_strdup(fallback);
+}
+
+static int
+int_or_fallback (GKeyFile *keyfile, const char *group,
+		const char *key, int fallback) {
+	g_autoptr(GError) local_error = NULL;
+	if (g_key_file_has_key(keyfile, group, key, &local_error))
+		return g_key_file_get_integer(keyfile, group, key, NULL);
+
+	return fallback;
+}
+
+static bool
+boolean_or_fallback (GKeyFile *keyfile, const char *group,
+		 const char *key, bool fallback, WacomStylus *stylus) {
+	g_autoptr(GError) error = NULL;
+	if (g_key_file_has_key(keyfile, group, key, &error)) {
+		return g_key_file_get_boolean(keyfile, group, key, NULL);
+	}
+	if (error && error->code == G_KEY_FILE_ERROR_INVALID_VALUE)
+		g_warning ("Stylus %s (%s) %s\n", stylus->name, group, error->message);
+
+	return fallback;
+}
+
+gchar **
+garray_to_string_list(GArray *array)
+{
+	gchar **values;
+	WacomStylusId *id;
+	char buffer[33];
+	bool ret;
+
+	if (array == NULL)
+		return NULL;
+
+	values = g_new (gchar *, array->len + 1);
```

Comment:
```suggestion
	values = g_new0 (gchar *, array->len + 1);
```
saves you manually null-terminating it a few lines below. Having said that - we're already using glib here so this could probably be something like:
```c

g_autoptr(GArray) *a = g_array_new(true, false, array->len);
for (...) {
   g_array_append_vals(...)
}
return g_array_steal(a);
```